Cancellation notification via email

aronerben opened the pull request:
Also add many many tests for sessions.

mabiede commented on pool/web/view/page/page_admin_session.ml:
```diff
@@ -817,3 +795,68 @@ let close
+       ; textarea_element
+           ~flash_fetcher
+           language
+           Pool_common.Message.Field.Reason
```
Message is defined at the beginning of the File ;-)

aronerben replied:
Thanks, done

mabiede commented on pool/web/view/page/page_admin_session.ml:
```diff
@@ -817,3 +795,68 @@ let close
+           language
+           Pool_common.Message.Field.Reason
+       ; span
+           Pool_common.
```
may use `let open Pool_common in` at the beginning of the function.

aronerben replied:
Thanks, I've done it everywhere in the file

mabiede commented on pool/app/session/session.ml:
```diff
@@ -11,3 +11,41 @@ let find_public_by_assignment = Repo.find_public_by_assignment
+  (* TODO [aerben] this language is probably wrong? dont take admin language,
+     take experiment language *)
+  let find = CCFun.flip (I18n.find_by_key tenant_db) language in
+  (* TODO [aerben] these should prob be exceptions... *)
```
handling those as results should be fine

aronerben replied:
https://www.joseferben.com/posts/error-handling-in-web-development/

@joseferben Makes the argument that certain problems should be exceptions. 

In this case, we have the case that cancellation text entries should be there guaranteed, as we have default values for all texts, should they not have provided custom translations. If those are missing, something has gone very wrong (db out of date or someone specifically deleted those rows). However, it is a bit weird to get a GitLab issue in that case through canary, so I'm kind of unsure. I changed the comment to be more general to reevaluate when to use exceptions. 

Maybe @joseferben could figure out a good concept for us?

joseferben replied:
In a way every bug should throw an exception, but not every exception is equal and different exceptions should be handled by different teams.

https://sentry.io/ works really well for me, I can highly recommend that.

aronerben replied:
Thanks for the input, I think this is a great solution.
- Treating all user errors (wrong input etc.) as `Result.t` and logging as INFO for us
- Treating all tenant-specific problems (like misconfiguration) as exceptions that trigger some warning (emails?) for tenants
- Treating programming errors or root problems as exceptions that trigger usual canary issues in GitLab

What do you think @mabiede @timohuber @asaurer (@joseferben )

asaurer replied:
@aronerben What kind of tenant-specific problems could arise? With the current setup, I still see our team responsible for them (SMTP settings, DB config, I18n, ...).

aronerben replied:
Will the tenant admins (not us) never configure something there? For example missing translations is something that sometimes happens currently if the DB is outdated

timohuber replied:
Tenants could activate a new language for example. Custom field names for example have to be provided if all active tenant languages. 

Issue 130 should cover that.

Not sure what other exceptions could happen due to operators actions.

joseferben replied:
Maybe any sort of exception (operator error or software bug/engineering error) could open an issue in Gitlab for now and a manual triage could assign issues to the right team? Once there is a sensible amount non-engineering issues coming in it could make sense to split the exceptions and let canary handle them differently?

aronerben commented on pool/web/handler/admin_session.ml:
```diff
@@ -305,7 +309,65 @@ let update_handler action req =
 let reschedule = update_handler `Reschedule
 
-let disabler req command ctor =
+(* TODO [aerben] if already canceled, allow uncancel *)
```
Currently discussing with @asaurer, I removed the TODO, we will put it into an issue if necessary.

aronerben commented:
@mabiede I
- added the checks so on-going sessions can be cancelled and after that they have to be closed
- added various tests
- extracted common functions into the model
- made the UI a bit more consistent when errors appear
- removed all redundant `create` functions
